whelk-core/src/main/groovy/whelk/filter/LinkFinder.groovy

```diff
@@ -112,72 +109,51 @@ class LinkFinder {
      * here.
      */
-    private void restoreNewCanonicalMainEntityUri(Map data) {
+    private static void restoreNewCanonicalMainEntityUri(Map data) {
         List items = data['@graph']
         if (items.size() < 2) {
             return
         }
         items[0][JsonLd.THING_KEY]['@id'] = items[1]['@id']
     }
 
-    private void replaceSameAsLinksWithPrimaries(Map data, List path = []) {
-        // If this is a link (an object containing _only_ an id)
-        String id = data.get("@id")
-        if (id != null && data.keySet().size() == 1) {
-            // Path to same form as in lddb__dependencies.relation
-            String normalizedPath = (path.take(2) == Document.recordPath
-                    ? [JsonLd.RECORD_KEY] + path.drop(2)
-                    : (path.take(2) == Document.thingPath ? path.drop(2) : path)
-            )
-                    .findAll { it instanceof String }
-                    .join('.')
-            if (RETAIN_NON_PRIMARY_IDS.contains(normalizedPath)) {
-                return
-            }
-            String primaryId = lookupPrimaryId(id, normalizedPath)
-            if (primaryId != null)
-                data.put("@id", primaryId)
-            return
-        }
-
-        // Keep looking for more links
-        for (Object key : data.keySet()) {
-
-            // sameAs objects are not links per se, and must not be replaced
-            String keyString = (String) key
-            if (keyString == "sameAs")
-                continue
-
-            Object value = data.get(key)
-
-            if (value instanceof List)
-                replaceSameAsLinksWithPrimaries((List) value, path + keyString)
-            if (value instanceof Map)
-                replaceSameAsLinksWithPrimaries((Map) value, path + keyString)
-        }
+    private void replaceSameAsLinksWithPrimaries(Document doc) {
+        // "Frame" temporarily to avoid cumbersome rewriting of property paths
+        Map thing = doc.getThing()
+        thing.put(RECORD_KEY, doc.getRecord())
+        replaceSameAsLinksWithPrimaries(thing, [])
+        // "Unframe"
+        thing.remove(RECORD_KEY)
     }
 
-    private void replaceSameAsLinksWithPrimaries(List data, List path) {
-        int idx = 0
-        for (Object element : data){
-            if (element instanceof List)
-                replaceSameAsLinksWithPrimaries( (List) element, path + idx )
-            else if (element instanceof Map)
-                replaceSameAsLinksWithPrimaries( (Map) element, path + idx )
-            idx += 1
+    private void replaceSameAsLinksWithPrimaries(data, List<String> propertyPath) {
+        if (data instanceof List) {
+            for (def element : data) {
+                replaceSameAsLinksWithPrimaries(data, propertyPath)
+            }
         }
-    }
+        if (data instanceof Map) {
+            // Keep original ID for "weak relations"
+            if (isLink(data) && !JsonLd.isWeak(propertyPath.join("."))) {
```

Comment:
Will not return here if it is weak link. 
i.e. the loop below will be run for weak links